### DISCUSSION TITLE
Failed to execute eth_sendTransaction requests using wallet connect v2 in our flutter dApp.

### DISCUSSION THREAD

**Discussion by michael16m2** [WalletConnect/walletconnect-monorepo#3094] [Developer Support]
We are currently in the process of migrating to wallet connect v2 and now got stuck trying to execute eth_sendTransaction requests, which was not a problem in case of v1. We use the Mumbai test network. Any advise or help in resolving it is highly appreciated:-) 

Implementation is based on the code snippets from https://github.com/WalletConnect/WalletConnectFlutterV2/tree/master. We manage to successfully create a session with MetaMask using SignClient. After that we try to execute a eth_sendTransaction as follows:

`
```
// Use Web3Dart to create the transaction parameters. 
// We call the safeTransfer method of a contract that deployed on mumbai.
var tx = Transaction.callContract(
      contract: contract,
      function: func!,  // safeTransfer
      parameters: params, // [ usrWalletAddress, receiverWalletAddress, tokenId]
      gasPrice: null,
      // gasPrice: EtherAmount.fromInt(EtherUnit.wei,
      //     1000000000), // If commented the value is taken automatically
      nonce: nonce,
    );

// create serialised json string for the request parameters
var reqParams = Params(from: usrWalletAddress, to: contractAddress, value: 0, data: tx.data);
String reqParamsJson = paramsToJson(reqParams);

launchUrlString(uri, mode: LaunchMode.externalApplication); // Activate MetaMask

final dynamic signResponse = await signClient.request(
      topic: session.topic,
      chainId: 'eip155:80001',
      request: SessionRequestParams(
        method: 'eth_sendTransaction',
        params: reqParamsJson,
      ),
    );

// Should we activate MetaMask app here?
// launchUrlString(uri, mode: LaunchMode.externalApplication); // Activate MetaMask

```

Executing the last step, signClient.request(), causes an json rpc exception error (or it may not return at all). 
A trace of the exception follows below:

_CastError (type 'Null' is not a subtype of type 'int' in type cast)
_column:
26
_errorMsg:
"type 'Null' is not a subtype of type 'int' in type cast"
_line:
10
_stackTrace:
#0      _$JsonRpcErrorFromJson (package:walletconnect_flutter_v2/apis/models/json_rpc_error.g.dart:10:26)
#1      new JsonRpcError.fromJson (package:walletconnect_flutter_v2/apis/models/json_rpc_error.dart:27:7)
#2      _$JsonRpcResponseFromJson (package:walletconnect_flutter_v2/apis/models/json_rpc_response.g.dart:18:26)
#3      new JsonRpcResponse.fromJson (package:walletconnect_flutter_v2/apis/models/json_rpc_response.dart:21:7)
#4      Pairing._onMessageEvent (package:walletconnect_flutter_v2/apis/core/pairing/pairing.dart:655:40)
<asynchronous suspension>
_url:
"package:walletconnect_flutter_v2/apis/models/json_rpc_error.g.dart"

```

**Comment by crypblizz8** [comment]
Passing this onto our team

**Reply by crypblizz8** [reply]
@ludowkm

**Reply by ludowkm** [reply]
hello @crypblizz8 , this is official library from wallet connect team, not us
Our official library is at https://pub.dev/packages/wallet_connect_v2
Currently wallet and dapp using our library still stable

**Reply by crypblizz8** [reply]
Sorry wrong tag! 

@Luzzotica

**Comment by Luzzotica** [comment]
I have created an example calling a smart contract function here:

https://github.com/WalletConnect/Web3ModalFlutter/blob/2981053456ecd3511ca2e3a7f0def7c75e22af62/example/sign/lib/utils/crypto/eip155.dart#L197

Please use that, and let me know if you have any issues.

**Reply by michael16m2** [reply]
@Luzzotica

Thank you very much for taking the effort. The example works fine. The example helped me to understand how to create the json transaction parameters to be passed to the request.  I can now also exclude few potential causes. I still have the same issue as mentioned earlier with our dApp however.

Some observations:
- I ran the example in a chrome browser on my Mac and use a Thrust wallet running on an Android phone. I adapted _testContractCall()_ in eip155.dart so that it uses our contract that we deployed on Polygon (calling the same function ). I do get the wallet UI to popup asking me for consent and the transaction appears successfully on Polyscan.
- Next I tried calling the contract in the same way and used the code from _testContractCall()_ in our dApp. The json parameter that I pass to the request is exactly the same as in the example. As mentioned earlier, the dApp is able to setup a session. But then executing the request using _SignClient_ or _Web3App_ does not have any effect. No exception raised this time.
-  I upgraded flutter (3.13.1) and WalletConnect v2 (2.1.4) with no effect. When creating an instance of _Web3App_ or _SignClient_ I now get an exception: _Failed host lookup: [relay.walletconnect.org](http://relay.walletconnect.org/)._ This error did not show before. Ignoring, the dApp still shows the same behaviour; setting up a session succeeds, but the request fails.

As I understand the main difference with the example is that the example uses services from walletconnect_modal to setup a session while I use _signClient.connect()_.  

Any suggestion that I could try next is highly appreciated!

**Reply by michael16m2** [reply]
@Luzzotica 
Unfortunately we still did not manage to execute transaction requests in case of V2. 
I created a simple app which uses the code from your example executing a request.  You can find it here: [wc2_test_request](https://github.com/michael16m2/wc2_test_request). It interacts with a contract that we deployed on Polygon. It was fine with v1.   
On IOS it *may* show the popup for consent, but that does not materialise into a successful transaction. On Android, it only works when I configure (pixel 6) to use private dns for some reason I do not understand.  
Could you perhaps have a quick look? I tried virtually everything I could think of but to no avail. Thanks!

**Reply by ludowkm** [reply]
@michael16m2 I think it would be better to use flutter wrapper https://pub.dev/packages/wallet_connect_v2 to reduce the risk, we can let WalletConnect team to maintain the native SDKs, it worked well at [Avacus app](https://avacus.cc)

**Reply by michael16m2** [reply]
@ludowkm Thanks for your suggestion, appreciate it. We actually did try it, but that also did not work. As it seems we are running out of options. As a temporary solution we consider to install a v1 server so that we can make some progress with our project again.

**Reply by ludowkm** [reply]
@michael16m2 It's stable at the moment with the version 1.0.7, could you share me which part does not work for you ?
It's impossible to install v1 server due to all DApps just support v2 only.

**Reply by rkovrigin** [reply]
Hello ludowkm, I work together with @michael16m2 
Today we tried that lib and we eventually could establish a session with TrustWallet, after that walletconnect_flutter_v2 also started to setup the session with TrustWallet.
Right now we have yet another problem with sending the transaction. 
Basically after "session.peer.redirect?.launch()" we are not redirected to our dApp.

I can provide an example code, that you can just run on your side, the only difference - we call a contract function.

**Reply by rkovrigin** [reply]
Please understand that this is an example code:
        
        import 'dart:io';
        
        import 'package:flutter/material.dart';
        import 'package:http/http.dart';
        import 'package:irdeto_dapp/common/metamask.dart';
        import 'package:irdeto_dapp/common/prefs.dart';
        import 'package:irdeto_dapp/pages/transaction_status.dart';
        import 'package:shared_preferences/shared_preferences.dart';
        import 'package:url_launcher/url_launcher.dart';
        import 'package:http/http.dart' as http;
        import 'package:url_launcher/url_launcher_string.dart';
        import 'dart:convert';
        import 'package:url_launcher/url_launcher.dart';
        
        import 'package:wallet_connect_v2/wallet_connect_v2.dart';
        import 'package:wallet_connect_v2/wallet_connect_v2.dart' as wc2;
        import 'package:web3dart/web3dart.dart';
        
        class TestSession extends StatefulWidget {
          @override
          State<TestSession> createState() => _TestSession();
        }
        
        class _TestSession extends State<TestSession> {
          static const projectId = '45caf086591c46efb6e9f19b6104d7e8';
        
          final _walletConnectV2Plugin = WalletConnectV2();
          final _walletMetadata = AppMetadata(
              name: 'Flutter Wallet',
              url: 'https://avacus.cc',
              description: 'Flutter Wallet by Avacus',
              icons: ['https://avacus.cc/apple-icon-180x180.png'],
              redirect: 'wcexample');
          final _uriController = TextEditingController();
          final List<Session> _sessions = [];
        
          late String _privateKey;
          late String _address;
          String? _dappTopic;
          String? _tempDappTopic;
          String? _uriDisplay;
        
          bool _isLoading = false;
          bool _isInitiated = false;
          bool _isForeground = true;
        
          @override
          void initState() {
            super.initState();
            _walletConnectV2Plugin.onConnectionStatus = (isConnected) async {
              debugPrint('---: CONNECTED: $isConnected');
              if (_isInitiated) {
                if (!isConnected && _isForeground) {
                  _walletConnectV2Plugin.connect();
                }
              } else {
                if (isConnected) {
                  setState(() {
                    _isLoading = false;
                    _isInitiated = true;
                  });
                  _refreshSessions();
                } else {
                  setState(() {
                    _isLoading = false;
                  });
                }
              }
            };
          }
        
          @override
          Widget build(BuildContext context) {
            return Scaffold(
                appBar: AppBar(
                  title: Text('Your transactions'),
                ),
                body: ListView(
                  children: [
                    ElevatedButton(
                      onPressed: () async {
                        createSession();
                        if (mounted) setState(() {});
                      },
                      child: Image.asset(
                        "assets/metamask.png",
                        scale: 10,
                      ),
                    ),
                    ElevatedButton(
                      onPressed: () async {
                        createSession();
                        if (mounted) setState(() {});
                      },
                      child: Text("Send transaction"),
                    ),
                  ],
                ));
          }
        
          void createSession() async {
            await _walletConnectV2Plugin.init(
                projectId: projectId, appMetadata: _walletMetadata);
            await _walletConnectV2Plugin.connect();
            final uri = await _walletConnectV2Plugin.createPair(namespaces: {
              'eip155': ProposalNamespace(chains: [
                'eip155:137'
              ], methods: [
                "eth_signTransaction",
                "eth_sendTransaction",
                "personal_sign",
                "eth_signTypedData"
              ], events: [])
            });
            if (uri == null) {
              print("URI IS EMPTY");
              return;
            } else {
              print("URI IS $uri");
              _uriDisplay = uri;
            }
        
            launchUrlString(uri, mode: LaunchMode.externalApplication);
        
            _tempDappTopic = uri.split('@')[0].split(':')[1];
            setState(() {});
          }
        
          Future _refreshSessions() async {
            try {
              final sessions = await _walletConnectV2Plugin.getActivatedSessions();
              if (sessions
                  .where((element) => element.topic == _dappTopic)
                  .toList()
                  .isEmpty) {
                _setDappTopic(_dappTopic = null);
              }
              _sessions.clear();
              _sessions.addAll(sessions);
              setState(() {});
            } catch (e) {
              _showDialog(child: Text('Refresh sessions error: ${e.toString()}'));
            }
          }
        
          Future _setDappTopic(String? topic) async {
            final sp = await SharedPreferences.getInstance();
            if (topic != null) {
              return sp.setString('dapp_topic', topic);
            }
            return sp.remove('dapp_topic');
          }
        
          Future _showDialog({required Widget child}) {
            return showDialog(
                context: context,
                builder: (context) {
                  return Dialog(
                    insetPadding: const EdgeInsets.symmetric(horizontal: 16),
                    backgroundColor: Colors.transparent,
                    child: Card(
                      clipBehavior: Clip.antiAlias,
                      color: Colors.white,
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(12)),
                      child: Container(
                        padding: const EdgeInsets.all(16),
                        child: child,
                      ),
                    ),
                  );
                });
          }
        
          void onSendTransactionTest() async {
            final session =
                _sessions.firstWhere((element) => element.topic == _dappTopic!);
        
            contractAddress = await getPrefsContractAddress();
            abiCode = await getAbiCode();
            EthereumAddress fromWallet =
                EthereumAddress.fromHex(await getPrefsUserWallet());
            EthereumAddress ethContractAddr = EthereumAddress.fromHex(contractAddress);
            final contract = DeployedContract(
                ContractAbi.fromJson(abiCode, "contract"), ethContractAddr);
        
            String chain = await getChainName();
            final client = Web3Client(rpcURLs[chain]!, Client());
            int nonce = await client
                .getTransactionCount(EthereumAddress.fromHex(fromWallet.toString()));
        
            if (contractAddress == "") {
              print("Contract address is not set!");
              return;
            }
        
            var params = [
              [3],
              4
            ];
            final ContractFunction? func = findFunctionInABI("redeemPokerHand", 2);
        
            var tx = Transaction.callContract(
              contract: contract,
              function: func!,
              parameters: params,
              // gasPrice: null,
              gasPrice: EtherAmount.fromInt(EtherUnit.wei,
                  1000000000), // If commented the value is taken automatically
              nonce: nonce,
            );
        
            var request = wc2.Request(
                method: 'eth_signTransaction',
                chainId: 'eip155:137',
                topic: _dappTopic!,
                params: [
                  {
                    "from": fromWallet.toString(),
                    "to": ethContractAddr.toString(),
                    "gas": "0x76c0", // 30400
                    "gasPrice": "0x9184e72a000", // 10000000000000
                    // "value": "0x9184e72a", // 2441406250
                    "data": tx.data
                  }
                ]);
            await _walletConnectV2Plugin.sendRequest(request: request);
            // launchUrlString(_uriDisplay!, mode: LaunchMode.externalApplication);
            // TODO: don't forget to check if where is the source come from to determine launch or not
            session.peer.redirect?.launch();
          }
        }
        
        extension StringExt on String {
          Future launch({int delayInMillis = 500}) async {
            try {
              await Future.delayed(Duration(milliseconds: delayInMillis));
              final uri = Uri.parse(contains(':') ? this : '$this:');
              if (startsWith('http')) {
                await _launchUniversalLink(uri);
              } else {
                await launchUrl(uri);
              }
            } catch (_) {}
          }
        
          Future _launchUniversalLink(Uri url) async {
            try {
              final bool nativeAppLaunchSucceeded = await launchUrl(
                url,
                mode: LaunchMode.externalNonBrowserApplication,
              );
              if (!nativeAppLaunchSucceeded) {
                return await launchUrl(
                  url,
                  mode: LaunchMode.externalApplication, // LaunchMode.inAppWebView
                );
              }
            } catch (_) {}
          }
        }
